Tweaks AI EMP resistances

atlantiscze opened the pull request:
- APCs with critical flag are now considerably more EMP resistant, failing for 1/10th of time and having better protection for the power cell. This prevents AIs from being one-shot killed by EMPing the core.
- AIs power restoration routine now tries to reboot the APC's failure timer when finished
- EMP still knocks out the AI for a moment, but isn't a guaranteed one hit kill due to power starvation.

**Updated:**
Okay, this spiraled out of control a bit. Well, it's done now, so here are the changes.

- Most of these changes are under the hood, and are barely noticeable for player. I will focus on the noticeable changes in this commit message.
- The power restoration routine is now rewritten to use (slightly) cleaner and easier to read code.
- All AI power interaction has been moved to separate file. One main has_power() proc has been created that replaced various other variants that sort of checked the power state. Power handling shifted to separate procs - one for power restoration routine, one for oxyloss (internal capacitor) handling and one for powersupply object synchronization and power usage updates.
- Considerably better synchronization between the AI and it's powersupply object. The powersupply should always be on AI's location, and it's active_power_usage should be managed by the procs the AI has.

**The AI now has set of power modes controlled by set of defines.**
- Charging: 70kW power usage, operates as usual and heals 2 oxyloss per tick. The define can be increased even more if we desire faster charging.
- Normal: 50kW power usage (as it was before this commit). Normal operation, the AI can work as usual. Usually this means the AI has external power from the APC. The AI can also force it's power on, at the cost of burning through it's remaining capacitor charge (oxyloss) very quickly. The AI can last for about one and half minute in this state, when running off internal capacitor.
- Restoration: 20kW power usage. The AI is blinded, can't control things and lacks most communication. This is automatically set when AI loses power, and comes with activation of special power restoration routine that tries to gain access to the APC. This routine reports status of the APC, and if possible, tries to reset it from EMP/misconfiguration to restore power to the AI.
- Low Power: 10kW power usage. The AI is blinded, can't control things and lacks most communication methods. This is set automatically if the AI's power restoration routine fails. It can also be set manually by the ai with the "Shutdown" command that's intended for situations where the AI wants to conserve remaining power to survive for a little longer. The AI can last for approximately 6 minutes in this state.

**New AI commands**
- "Shutdown" - Allows the AI to shut itself down, reducing power usage to 10kW. This however prevents power restoration routine from running. The AI behaves as if it was without power, except that it doesn't take oxyloss damage (assuming the APC still provides the 10kW necessary for the AI to operate). It may be used again to cancel the shutdown, returning to normal mode. Intention of this is to provide an emergency option for the AI, when it's power is very low and when help needs time to arrive. By sacrificing communication/control/camera capabilities the AI can multiply it's remaining APC battery life up to five times.
- "Toggle Power Override" - The AI can enable this to remain online despite the lack of external APC power. It however causes much more oxyloss per second. The idea behind this is to allow the AI to send a quick yell for help over the radio, or to allow the AI to correct it's own SMES misconfiguration or other problem that can be fixed quickly. I once again point out that the AI lasts for less than two minutes in this mode, so it's not advised to keep it on at all times, and only use it sparingly.

**Note**
I have noticed a minor bug, where the AI's blind overlay when powerless flickers on and off. I have been unable to determine the cause, and i am fairly certain it was there even before, and thus it's not linked to the changes in this PR. Most likely a 510-related issue. The effect is visual only - the AI can't interact with any machines, etc. despite being able to briefly see them. I guess it's not as big problem for now.

mwerezak commented on code/modules/power/apc.dm:
```diff
@@ -1168,9 +1169,18 @@ obj/machinery/power/apc/proc/autoset(var/val, var/on)
+	if(is_critical)
+		// Critical APCs are considered EMP shielded and will be offline only for about half minute. Prevents AIs being one-shot disabled by EMP strike.
+		// Critical APCs are also more resilient to cell corruption/power drain.
+		energy_fail(round((round(rand(240, 360) / severity) / CRITICAL_APC_EMP_PROTECTION)))
```
Unnecessary amounts of rounding.

If you move the rounding into energy_fail() you can also avoid some copypasta here too.

atlantiscze replied:
Good idea. Changed.

mwerezak commented on code/modules/power/apc.dm:
```diff
@@ -1168,9 +1169,18 @@ obj/machinery/power/apc/proc/autoset(var/val, var/on)
+		// Critical APCs are also more resilient to cell corruption/power drain.
+		energy_fail(rand(240, 360) / severity / CRITICAL_APC_EMP_PROTECTION)
+		if(cell)
+			cell.emp_act(severity+2)
```
You didn't mention this change, is it really needed?

atlantiscze replied:
It is mentioned:    .... and having better protection from the power cell. ...

I admit, i made a type, should've been for the power cell, rather than from.

mwerezak replied:
Yeah, it's a buff, but not a huge problem I suppose.

mwerezak commented on code/modules/mob/living/silicon/ai/life.dm:
```diff
@@ -116,6 +116,7 @@
 									sleep(2)
 									theAPC.operating = 1
 									theAPC.equipment = 3
+									theAPC.failure_timer = 0
```
Can you do me a favour and move everything from line 63 included into a separate proc?

atlantiscze replied:
Sure

atlantiscze commented:
The PR has been updated. I have sort of overhauled AI power interaction almost completely.

mwerezak commented on code/modules/mob/living/silicon/ai/power.dm:
```diff
@@ -0,0 +1,220 @@
+// Handles all necessary power checks: Area power, Intellicard and Malf AI APU power and manual override.
+/mob/living/silicon/ai/proc/has_power(var/respect_override = 1)
+	var/area/current_area = get_area(src)
+	if(current_area && (current_area.power_equip || current_area.requires_power == 0))
```
It would make for cleaner integration with machine power if `psupply.powered()` was used here.

PsiOmegaDelta commented:
All these if-statements could be changed into a more maintainable (datum) state machine.

atlantiscze commented:
Now using powersupply's powered() instead of getting area power state directly.

Fixed a minor bug where using shutdown command once power restoration routine started resulted in reduced power usage, despite having active power restoration routine.

I've also replaced hardcoded numbers with defines. Is this what you meant, @PsiOmegaDelta ? If not, please clarify it a bit more.

mwerezak commented on code/modules/mob/living/silicon/ai/power.dm:
```diff
@@ -0,0 +1,214 @@
+/mob/living/silicon/ai/proc/update_power_usage()
+	var/newusage = calculate_power_usage()
+	newusage *= AI_POWERUSAGE_OXYLOSS_TO_WATTS_MULTIPLIER
+	if(psupply)
```
Is it ever valid for an AI not not have a psupply?

atlantiscze replied:
Extra sanity doesn't hurt, i guess. It's not a code that would be called hundreds of time every tick so the performance difference is very close to zero.

mwerezak replied:
Fair enough

mwerezak commented on code/modules/mob/living/silicon/ai/power.dm:
```diff
@@ -0,0 +1,214 @@
+/obj/machinery/ai_powersupply/New(var/mob/living/silicon/ai/ai=null)
+	powered_ai = ai
+	powered_ai.psupply = src
+	forceMove(powered_ai.loc)
```
Huh, didn't realize that the power supply machine wasn't inside the AI.

It probably would be better if it was inside the AI, then you wouldn't need to worry about syncing their locs.

Have it start inside the given AI, then add the following.
```
/obj/machinery/ai_powersupply/powered(var/chan = -1, var/area/check_area = null)
    return ..(chan, powered_ai.loc.loc)
```

mwerezak replied:
I can't remember though if machines need to be on a turf though for `auto_use_power()` to work.

atlantiscze replied:
wouldn't get_area(powered_ai) be better? AI's loc is not guaranteed to be a turf, AI can be held inside an item too, for example.

mwerezak replied:
I was just following what the base proc did. get_area() works too.

mwerezak commented on code/modules/mob/living/silicon/ai/power.dm:
```diff
@@ -0,0 +1,214 @@
+	if(powered_ai.APU_power || powered_ai.admin_powered || istype(powered_ai.loc, /obj/item/))
+		use_power = 0
+		return
+	use_power = 2
```
Honestly there really is no reason to process any of this stuff. 

* Have AI set `psupply.use_power = 0` on death
* `powered_ai.psupply != src` should never be true
* `forceMove(powered_ai.loc)` is unnecessary once you move the power supply into the AI mob
* Set `psupply.use_power` when `APU_power` or `admin_powered` are set, or when the AI is carded.

mwerezak commented:
@atlantiscze you're using a state machine around the `aiRestorePowerRoutine` var right? I think Psi wants you to make your state machine a bit clearer, perhaps set aside a place where you explicitly define the state transitions and put the actions of each state under a switch for consistency.

Also it might be cleaner to integrate the self power down/special admin up states into the state machine instead of having them just hanging around.

PsiOmegaDelta commented:
This is what I mean by state machine: https://en.wikipedia.org/wiki/State_machine. In the world of BYOND the OOP approach would basically be a set of datums each implementing the same proc, with a way to transition between them.
It's not something I'm expecting tho, would just clarify things.

atlantiscze commented:
Tweaked a bit. The location synchronizing code is gone, and powersupply is inside the AI now. I've also replaced the code in process(), right now it seems to be the best i can do without going through pretty much any case that may cause AI's power usage to change.

As for the state machine thing, that's kind of out of the scope of this PR. The system i used right now works well, the transitions are pretty much clear (the power restoration advances to next stage every iteration, unless the current stage is -1 (AI_RESTOREPOWER_FAILED)). The only thing that directly alters the aiRestorePowerRoutine variable is the power restoration proc, and the piece of code under ifs that triggers it in AI's Life().

mwerezak commented:
:+1:

PsiOmegaDelta commented on code/modules/mob/living/silicon/ai/ai.dm:
```diff
@@ -262,55 +267,16 @@ var/list/ai_verbs_default = list(
-		aiPDA.set_owner_rank_job(pickedName, "AI")
+		aiPDA.ownjob = "AI"
+		aiPDA.owner = pickedName
+		aiPDA.name = pickedName + " (" + aiPDA.ownjob + ")"
```
Do use the ```set_owner_rank_job()``` proc that ensures the PDA is setup proper.

atlantiscze replied:
Seems this may have been lost in a merge. Will fix.

atlantiscze commented:
A dev might have to poke travis or something:

No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated

PsiOmegaDelta commented:
Torch is becoming infamous for stalling DD.
